### ISSUE TITLE
for loop in map_fn within tf.function is very slow compared to using while_loop counterpart

### ISSUE THREAD

**Issue by henrysky** [tensorflow/tensorflow#40517]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): N/A
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Win10 2004 x64
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.2.0
- Python version: 3.7.7
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: 10.1, 7,6
- GPU model and memory: 1650Ti

**Describe the current behavior**

Code 1 runs much slow than Code 2 (71s seconds vs 11s), but from my understanding in Tensorflow documentation, Code 1 and Code 2 should be effectively the same (tf.function will turn the for loop into `while_loop` if the loop use Tensor but apparently not the case here). Moreover Code 2 will run much faster (~1s) on CPU (9750H, 1650Ti), not sure if that is expected tho. Code 3 using for loop and for loop to replace `map_fn` completed in 164s which is the slowest among the three which is expected.

*Update:* This behavior is not observed in tf2.1.0, code 1/2/3 completed in similar amount of time (11ish seconds)

**Describe the expected behavior**

Code 1 and Code 2 runs as fast as each other.

**Standalone code to reproduce the issue**

Code 1:
```python3
import time
import tensorflow as tf

@tf.function
def odeint_external(tensor):
    finished_user_t_ii = 0
    x = tensor[0]
    t = tensor[1]
    
    # array to store the result
    result = tf.TensorArray(dtype=tf.float32, size=t.shape[0])
    result = result.write(0, x)

    i = tf.constant(0, dtype=tf.int32)
    for _t in t[1:]:
        i = i+1
        result = result.write(i, tf.stack([tf.cos(_t), tf.sin(_t)]))

    return result.stack(), t


@tf.function
def parallelized_map_fn(tensor):
    return tf.map_fn(odeint_external, tensor)


# warm up
result = parallelized_map_fn((tf.random.normal((50, 2), 0, 1), tf.random.normal((50, 10000), 0, 1)))

start_t = time.time()
result = parallelized_map_fn((tf.random.normal((50, 2), 0, 1), tf.random.normal((50, 10000), 0, 1)))
print(time.time()-start_t)
```

Code 2:
```python3
import time
import tensorflow as tf

@tf.function
def odeint_external(tensor):
    finished_user_t_ii = 0
    x = tensor[0]
    t = tensor[1]
    
    # array to store the result
    result = tf.TensorArray(dtype=tf.float32, size=t.shape[0])
    result = result.write(0, x)

    i = tf.constant(1, dtype=tf.int32)
    cond = lambda i, _t, r: i < 9999
    body = lambda i, _t, r: (i+1, _t, r.write(i, tf.stack([tf.cos(_t[i]), tf.sin(_t[i])])))
    result = tf.while_loop(cond=cond, body=body, loop_vars=(i, t[1:], result))
    
    return result[2].stack(), t

@tf.function
def parallelized_map_fn(tensor):
    return tf.map_fn(odeint_external, tensor, parallel_iterations=1)

# warm up
result = parallelized_map_fn((tf.random.normal((50, 2), 0, 1), tf.random.normal((50, 10000), 0, 1)))

start_t = time.time()
result = parallelized_map_fn((tf.random.normal((50, 2), 0, 1), tf.random.normal((50, 10000), 0, 1)))
print(time.time()-start_t)
```

Code 3:
```python
import time
import tensorflow as tf

@tf.function
def odeint_external(tensor):
    finished_user_t_ii = 0
    x = tensor[0]
    t = tensor[1]
    
    # array to store the result
    result = tf.TensorArray(dtype=tf.float32, size=t.shape[0])
    result = result.write(0, x)

    i = tf.constant(0, dtype=tf.int32)
    for _t in t[1:]:
        i = i+1
        result = result.write(i, tf.stack([tf.cos(_t), tf.sin(_t)]))
    
    return result.stack(), t

x = tf.random.normal((50, 2), 0, 1)
t = tf.random.normal((50, 10000), 0, 1)

# warm up
result = [odeint_external((_x, _t)) for _x, _t in zip(x, t)]

start_t = time.time()
result = [odeint_external((_x, _t)) for _x, _t in zip(x, t)]
print(time.time()-start_t)
```

**Other info / logs** Include any logs or source code that would be helpful to
N/A

**Comment by ravikyram** [comment]
I have tried in colab with TF version 2.2, nightly versions and was able to reproduce the issue .Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/51b9c8e10a6bb308a3336943defb457d/untitled35.ipynb).Thanks!

**Comment by mdanatg** [comment]
This looks like a regression - certain ops that the for loop uses don't seem to be optimized properly. We'll look into that.

For a temporary workaround, writing the loop like so should avoid the slow ops:

```
    i = tf.constant(1, dtype=tf.int32)
    while i < 9999:
        result = result.write(i, tf.stack([tf.cos(t[i]), tf.sin(t[i])]))
        i = i+1
```

**Comment by mdanatg** [comment]
Looks like the workaround only fixes the difference partially - there is still a pair of logical_and/logical_or ops which are slow on GPU and hurt performance [1]. This will be fixed in TF 2.4.

https://github.com/tensorflow/tensorflow/blob/4e73f5f852941754f1f4a2880bf92d5d2b26e51b/tensorflow/python/autograph/operators/control_flow.py#L525
